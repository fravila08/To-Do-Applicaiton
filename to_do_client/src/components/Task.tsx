import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import { ITask } from "../App";
import Form from "react-bootstrap/Form";
import axios from "axios";
import done from "../assets/done.png";
import pending from "../assets/pending.png";

export interface TaskProps {
  task: ITask;
  selectedTasks: number[];
  setSelectedTasks: (selectedTasks: number[]) => void;
  allTasks: ITask[];
  setAllTasks: (allTasks: ITask[]) => void;
}

export const changeTaskStatus = async (id: number) => {
  try {
    let response = await axios.put(`task/${id}/`);
    return response.data.changed;
  } catch (err) {
    alert(err);
    return false;
  }
};

export const Task: React.FC<TaskProps> = ({
  task,
  setSelectedTasks,
  selectedTasks,
  allTasks,
  setAllTasks,
}) => {
  const changeStatus = async (clicked: boolean, taskToChange: ITask) => {
    let response = await changeTaskStatus(taskToChange["id"]);
    if (response) {
      setAllTasks(allTasks.filter((task) => task !== taskToChange));
      taskToChange.completed = clicked;
      setAllTasks([...allTasks]);
    }
  };

  return (
    <Row className="task">
      <Col>
        <Form.Check
          type="checkbox"
          id={`taskSelectedBtn${task.id}`}
          onChange={(e) =>
            e.target.checked
              ? setSelectedTasks([...selectedTasks, task.id])
              : setSelectedTasks(selectedTasks.filter((id) => id !== task.id))
          }
        />
      </Col>
      <Col id={`task${task.id}`} xs={7} className="taskTitle">
        {task.title}
      </Col>
      <Col className="checkHolder" xs={2}>
        <Form.Check
          id={`taskCheck${task.id}`}
          type="checkbox"
          name="checkbox"
          className="check"
          checked={task.completed}
          onChange={(e) => changeStatus(e.target.checked, task)}
        />
        <Form.Label for="checkbox">
          <img className="checkImg" src={task.completed ? done : pending} />
        </Form.Label>
      </Col>
    </Row>
  );
};
