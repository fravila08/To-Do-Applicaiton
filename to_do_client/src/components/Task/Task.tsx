import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Form from 'react-bootstrap/Form'
import { useEffect, useState } from 'react'
import { Tasks } from '../../App'

interface IdSelectionProps{
    selected: number[];
    setSelected: (selected: number[]) => void;
    task: Tasks;
}

// export function addToSelection(selected:number[],id:number, checked:boolean){
//     console.log("hello")
//     let mySelected=selected
//     if(checked){
//         mySelected.push(Number(id))
//     }
//     else{
//         let newSelected:number[]=[]
//         for(let i=0; i< mySelected.length; i++){
//             if(mySelected[i]!==id){
//                 newSelected.push(mySelected[i])
//             }
//         }
//         mySelected=newSelected
//     }
//     console.log(mySelected,'s')
//     return mySelected
// }


export const Task: React.FC<IdSelectionProps>=({setSelected, selected, task})=>{
    const [mySelected, setMySelected]=useState<number[]>(selected)
    
    useEffect(()=>{
        setMySelected(selected)
    },[selected])


    return(
        <Row style={{backgroundColor:"grey", border:"black 2px solid"}}>
            <Col xs={9}>
                {task.title}
            </Col>
            <Col xs={2}>
                <Form.Check type="checkbox" label="select" 
                onChange={(e)=>{
                    if(e.target.checked){
                    setSelected([...selected, task.id])
                    }
                    else{ setSelected(selected.filter(item=> item !== task.id))}
                    }} 
                
                />
            </Col>
        </Row>
    )
}