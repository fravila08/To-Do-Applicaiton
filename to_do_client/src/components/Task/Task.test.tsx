import { Task} from "./Task";
import {addToSelection} from "./Task"
import { useState } from "react";

import { describe, expect, it, vi } from "vitest";
import TestRenderer from 'react-test-renderer';

describe("This will control the array of selected id's",()=>{
    it("will take in an id, whether a task is checked, and the useState functions for selectedIDs",()=>{
        let selected=[]
        const setSelected=(v)=>{selected=v}
        let checked=true
        let id=1
        expect(addToSelection(selected, setSelected, id, checked)).toEqual([1])
    })
})

describe("This will control the array of selected id's",()=>{
    it("will take in an id, whether a task is checked, and the useState functions for selectedIDs",()=>{
        let selected=[1]
        const setSelected=(v)=>{selected=v}
        let checked=false
        let id=1
        expect(addToSelection(selected, setSelected, id, checked)).toEqual([])
    })
})

describe("Will take a snapshot of the current Task compenent",()=>{
    it("will create a snapshot",()=>{
        let selected=[]
        const setSelected=(v)=>{selected=v}
        const header= TestRenderer.create(<Task 
            setSelected={setSelected} 
            selected={selected} 
            task={{"id":1,"title":"Style"}}
          />)
        expect(header).toMatchSnapshot()
    })
})