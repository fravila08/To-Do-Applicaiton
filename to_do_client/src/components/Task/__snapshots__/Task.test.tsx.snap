// Vitest Snapshot v1

exports[`Will take a snapshot of the current Task compenent > will create a snapshot 1`] = `
<div
  className="row"
  style={
    {
      "backgroundColor": "grey",
      "border": "black 2px solid",
    }
  }
>
  <div
    className="col-9"
  >
    Style
  </div>
  <div
    className="col-2"
  >
    <div
      className="form-check"
    >
      <input
        className="form-check-input"
        disabled={false}
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        title=""
      >
        select
      </label>
    </div>
  </div>
</div>
`;
